### ISSUE TITLE
Question: modelling shipments with lots of possible time-frames

### ISSUE THREAD

**Issue by fbaierl** [VROOM-Project/vroom#679]
Hello,

I have been wondering if it would pose a big performance-problem to add multiple (many) possible time-frames for a shipment. 
My use case is the following:

- we are transporting people
- we can change the start-time for a transport +/- 15 minutes
- drivers have to take breaks (30 minutes a day)
- we wouldn't want drivers to take their break while people are being transportet

This leads to a lot of "shipments" being defined like this:

ordered: 8:00 o'clock, 20 minute drive
-> start between 7:45 and 8:15 (8:00 - 0:15; 8:00 + 0:15)
-> end between 8:05 and 8:35 (8:20 - 0:15; 8:20 + 0:15)

So, for a 20 minute drive, we end up with a maximum time-window of 50 minutes - for that reason vroom often schedules the break while we are transporting people. That is something we do not want.

My proposed workaround would be to define multiple time-windows for the shipment, such as:

7:45 -> 8:05
7:50 -> 8:10
7:55 -> 8:15
...

Would that introduce a huge additional performance loss? Or is there a better way to model our problem? 

BTW: thank you very much for making such a powerful library open-source

**Comment by jcoupey** [comment]
> My proposed workaround would be to define multiple time-windows for the shipment

Do you mean you'd apply that same set of time windows for both pickup and delivery in your shipments?

**Comment by fbaierl** [comment]
I just realized that my solution doesn't make any sense, since shipments don't have time-windows, only the pickup and delivery shipment steps.

So, at the moment it is not possible at all to model flexbile time-windows for pickup/delivery (+/- 15 minutes) with a defined max. time between pickup and delivery?

To reiterate my original problem: I want **the start (pickup)** of a shipment to be flexible, without giving the vehicle leeway to lengthen the time of the drive from pickup to delivery. So e.g. a vehicle can start between 7:45 and 8:15, but should drive to the delivery location as fast as possible.

 Is there any solution for that currently?

**Comment by jcoupey** [comment]
> a defined max. time between pickup and delivery

No, there is no way to master this constraint right now except indirectly by the amplitude between delivery deadline and pickup available date. I'm inclined to think that it would be hard to add such a constraint with the current TW modelling setup we have.

You're trying to make the "no break with passenger" requirement a consequence of other constraints. While this kind of workaround works in many situations, I think here it would probably be easier to stick to the actual requirement. This would mean adjusting the validity checks we do internally to discard all solutions where a break occur at a point where a vehicle is not empty.

**Comment by fbaierl** [comment]
> This would mean adjusting the validity checks we do internally to discard all solutions where a break occur at a point where a vehicle is not empty.

That is probably something we need to do then, yes. Would `TWRoute::is_valid_addition_for_tw` be the right place for something like that?

Unfortunately though, I don't think that would suffice, because even without a break, a vehicle can still use the 'extra time' given to do other tasks.

**Comment by jcoupey** [comment]
Yes, the whole validity logic related to time constraints is in `TWRoute::is_valid_addition_for_tw`.

**Comment by fbaierl** [comment]
Alright, I'll close this issue then. Thanks for the help.
